staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storageversion
+
+import (
+	"context"
+
+	apiextensionshelpers "k8s.io/apiextensions-apiserver/pkg/apihelpers"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	apiserverinternalv1alpha1 "k8s.io/api/apiserverinternal/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	genericstorageversion "k8s.io/apiserver/pkg/storageversion"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// updateQueueBufferSize is the channel buffer size for each
+	// StorageVersion updateQueue. Since the storage version manager keeps
+	// one updateQueue for each CRD UID, this means each CRD may have at
+	// most 10 pending storage version updates in the queue. When a queue
+	// is full, new storage version updates will be dropped.
+	updateQueueBufferSize = 10
+)
+
+// Manager provides methods for updating StorageVersion for CRDs. It does
+// goroutine management to allow each CRD having its own updates running in the
+// background and not blocking each other, while making sure the updates for the
+// same CRD happen in the same order as the methods get invoked.
+// NOTE: the caller should hold a lock while calling the methods, supposedly the
+// customStorageLock, to make sure the storage version update is in sync with
+// the CRD underlying storage change.
```

Comment:
this is not completely clear here, especially because `UpdateStorageVersionFor` just queues and returns.

Reply:
moved lock-related comments to crdHandler. The manager doesn't care about the lock